src/compute/arithmetics/timestamp.rs

```diff
@@ -0,0 +1,351 @@
+// Licensed to the Apache Software Foundation (ASF) under one
+// or more contributor license agreements.  See the NOTICE file
+// distributed with this work for additional information
+// regarding copyright ownership.  The ASF licenses this file
+// to you under the Apache License, Version 2.0 (the
+// "License"); you may not use this file except in compliance
+// with the License.  You may obtain a copy of the License at
+//
+//   http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing,
+// software distributed under the License is distributed on an
+// "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY
+// KIND, either express or implied.  See the License for the
+// specific language governing permissions and limitations
+// under the License.
+
+//! Defines the arithmetic kernels for Timestamp and Duration.
+//! For the purposes of Arrow Implementations, adding this value to a Timestamp
+//! ("t1") naively (i.e. simply summing the two number) is acceptable even
+//! though in some cases the resulting Timestamp (t2) would not account for
+//! leap-seconds during the elapsed time between "t1" and "t2".  Similarly,
+//! representing the difference between two Unix timestamp is acceptable, but
+//! would yield a value that is possibly a few seconds off from the true elapsed
+//! time.
+
+use crate::{
+    array::{Array, PrimitiveArray},
+    compute::arity::unchecked_binary,
+    datatypes::{DataType, TimeUnit},
+    error::{ArrowError, Result},
+};
+
+/// Calculates the scale factor between two TimeUnits. The function returns the
+/// scale that should multiply the TimeUnit "b" to have the same time scale as
+/// the TimeUnit "a".
+fn timeunit_scale(a: &TimeUnit, b: &TimeUnit) -> f64 {
```

Comment:
What do you think about moving this to the `temporal_conversions` module? Seems related.